contracts/Gauge.sol

```diff
@@ -383,162 +450,246 @@ contract Gauge is BaseGauge, IGauge {
      *  @param _addr address to change approval for
-     *  @param _canDeposit can deposit
      *  @param _canClaim can deposit
      *  @return true
      */
     function setApprovals(
         address _addr,
-        bool _canDeposit,
         bool _canClaim,
         bool _canLock
     ) external returns (bool) {
-        approvedTo[_addr][msg.sender].deposit = _canDeposit;
         approvedTo[_addr][msg.sender].claim = _canClaim;
         approvedTo[_addr][msg.sender].lock = _canLock;
 
         return true;
     }
 
-    /** @notice withdraw vault token from the gauge
+    /** @notice Burns shares from owner and sends exactly assets of underlying tokens to receiver.
      *  @dev This call updates claimable rewards
-     *  @param _amount amount to withdraw
-     *  @param _claim claim veYFI and additional reward
+     *  @param _assets amount to withdraw
+     *  @param _receiver account that will recieve the shares
+     *  @param _owner shares will be taken from account
      *  @param _lock should the claimed rewards be locked in veYFI for the user
-     *  @return true
+     *  @param _claim claim veYFI and additional reward
+     *  @return amount of shares withdrawn
      */
     function withdraw(
-        uint256 _amount,
+        uint256 _assets,
+        address _receiver,
+        address _owner,
         bool _claim,
         bool _lock
-    ) public updateReward(msg.sender) returns (bool) {
-        require(_amount != 0, "RewardPool : Cannot withdraw 0");
-        Balance storage balance = _balances[msg.sender];
-        require(
-            balance.lastDeposit < block.number,
-            "no withdraw on the deposit block"
-        );
-
-        //also withdraw from linked rewards
-        uint256 length = extraRewards.length;
-        for (uint256 i = 0; i < length; ++i) {
-            IExtraReward(extraRewards[i]).rewardCheckpoint(msg.sender);
-        }
-
-        _totalSupply -= _amount;
-        uint256 newBalance = balance.realBalance - _amount;
-        balance.realBalance = newBalance;
-        uint256 boostedBalance = _boostedBalanceOf(msg.sender, newBalance);
-        balance.boostedBalance = boostedBalance;
-        emit BoostedBalanceUpdated(msg.sender, boostedBalance);
-
-        if (_claim) {
-            _getReward(msg.sender, _lock, true);
-        }
-
-        stakingToken.safeTransfer(msg.sender, _amount);
-        emit Withdrawn(msg.sender, _amount);
+    ) external returns (uint256) {
+        return _withdraw(_assets, _receiver, _owner, _claim, _lock);
+    }
 
-        return true;
+    /** @notice Burns shares from owner and sends exactly assets of underlying tokens to receiver.
+     *  @dev This call updates claimable rewards
+     *  @param _assets amount to withdraw
+     *  @param _receiver account that will recieve the shares
+     *  @param _owner shares will be taken from account
+     *  @return amount of shares withdrawn
+     */
+    function withdraw(
```

Comment:
I would add `_` as prefix (internal function).

Reply:
This should be external